File: kickstart/2022/RoundA/challenge_nine.cpp
#include <bits/stdc++.h>

using namespace std;

int which_digit_to_insert(string &s)
{
    int sum = 0;
    for (int i = 0; i < s.size(); ++i)
    {
        sum += int(s[i] - '0');
    }
    if (sum % 9 != 0)
    {
        return 9 - sum % 9;
    }
    else
    {
        return 0;
    }
}

vector<int> get_digits(string &s, int L)
{
    vector<int> digits;
    for (int i = 0; i < L; ++i)
    {
        digits.push_back(int(s[i] - '0'));
    }
    return digits;
}

int where_to_insert(string &s, int d)
{
    int k = 0;
    int L = s.size();
    vector<int> digits = get_digits(s, L);
    for (int i = 0; i < L; ++i)
    {
        if (digits[i] > d)
        {
            k = L - i;
            break;
        }
    }
    if (d == 0 && k == L)
    {
        k -= 1;
    }
    return k;
}

string insert_digit_str(string &s, int d, int k)
{
    string before = s.substr(0, s.size() - k);
    string after = s.substr(s.size() - k);
    return before + to_string(d) + after;
}

int main()
{
    int T;
    cin >> T;
    for (int t = 1; t < T + 1; ++t)
    {
        string s;
        cin >> s;
        int d = which_digit_to_insert(s);
        int k = where_to_insert(s, d);
        cout << "Case #" << t << ": " << insert_digit_str(s, d, k) << endl;
    }
}

File: kickstart/2022/RoundA/challenge_nine.py
def solution1():
    N = input().rstrip()
    org_N = int(N)
    N = list(N)
    cands = []
    for i in range(len(N), -1, -1):
        for j in range(10):
            n = N[:i] + [str(j)] + N[i:]
            n = int(''.join(n))
            if n != org_N and n % 9 == 0:
                cands.append(n)
    print('Case #{}: {}'.format(t, min(cands)))


def insert_digit(N, d, k):
    return (N - N % 10**k) * 10 + d * 10**k + (N % 10**k)


def insert_digit_with_slice(N, d, k):
    N = list(str(N))
    n = N[:k] + [str(d)] + N[k:]
    return int(''.join(n))


def solution2():
    import math
    N = int(input().rstrip())
    L = len(str(N))
    cands = []
    for i in range(L + 1):
        for j in range(10):
            n = insert_digit(N, j, i)
            if n != N and n % 9 == 0:
                cands.append(n)
    print('Case #{}: {}'.format(t, min(cands)))


def which_digit_to_insert(N):
    sum_n = sum(list(map(int, list(str(N)))))
    return 9 - sum_n % 9 if sum_n % 9 != 0 else 0


def where_to_insert(N, d):
    def get_digits(N):
        if N < 10:
            return [N]
        ret = get_digits(N // 10)
        ret.append(N % 10)
        return ret

    def get_digits_iter(N, L):
        digits = []
        rem = N
        while L >= 1:
            d = rem // 10**(L-1)
            rem = rem % 10**(L-1)
            digits.append(d)
            L = L - 1
        return digits
    k = 0
    L = len(str(N))
    digits = get_digits_iter(N, L)
    for i, digit in enumerate(digits):
        if digit > d:
            break
        k += 1
    return k + 1 if d == 0 and k == 0 else k


def solution3():
    import math
    N = int(input().rstrip())
    d = which_digit_to_insert(N)
    k = where_to_insert(N, d)
    print('Case #{}: {}'.format(t, insert_digit_with_slice(N, d, k)))


T = int(input())
for t in range(1, T + 1):
    solution3()
